mender-auth/client/client.cpp

```diff
@@ -0,0 +1,210 @@
+	if (!expected_public_key) {
+		return expected_public_key.error();
+	}
+	request_body_map.insert({"pubkey", expected_public_key.value()});
+
+	auto expected_request_body = json::Dump(request_body_map);
+	if (!expected_request_body) {
+		return expected_request_body.error();
+	}
+	auto request_body = expected_request_body.value();
+
+	// Sign the body
+	auto expected_signature =
+		crypto::Sign(private_key_path, common::ByteVectorFromString(request_body));
+	if (!expected_signature) {
+		return expected_signature.error();
+	}
+	auto signature = expected_signature.value();
+
+	auto whole_url = path::Join(server_url, request_uri);
+
+	auto req = make_shared<http::OutgoingRequest>();
+	req->SetMethod(http::Method::POST);
+	req->SetAddress(whole_url);
+	req->SetHeader("Content-Type", "application/json");
+	req->SetHeader("Content-Length", to_string(request_body.size()));
+	req->SetHeader("Accept", "application/json");
+	req->SetHeader("X-MEN-Signature", signature);
+	req->SetHeader("Authorization", "API_KEY");
+
+	req->SetBodyGenerator([request_body]() -> io::ExpectedReaderPtr {
+		return make_shared<io::StringReader>(request_body);
+	});
+
+	auto received_body = make_shared<vector<uint8_t>>();
+
+	return client.AsyncCall(
+		req,
+		[received_body, api_handler](http::ExpectedIncomingResponsePtr exp_resp) {
+			if (!exp_resp) {
+				mlog::Error("Request failed: " + exp_resp.error().message);
+				api_handler(expected::unexpected(exp_resp.error()));
+				return;
+			}
+			auto resp = exp_resp.value();
+
+			auto body_writer = make_shared<io::ByteWriter>(received_body);
+			body_writer->SetUnlimited(true);
+			resp->SetBodyWriter(body_writer);
+
+			mlog::Debug("Received response header value:");
+			mlog::Debug("Status code:" + to_string(resp->GetStatusCode()));
+			mlog::Debug("Status message: " + resp->GetStatusMessage());
+		},
+		[received_body, api_handler](http::ExpectedIncomingResponsePtr exp_resp) {
+			if (!exp_resp) {
+				mlog::Error("Request failed: " + exp_resp.error().message);
+				api_handler(expected::unexpected(exp_resp.error()));
+				return;
+			}
+			auto resp = exp_resp.value();
+
+			string response_body = common::StringFromByteVector(*(received_body.get()));
+
+			switch (resp->GetStatusCode()) {
+			case http::StatusOK:
+				api_handler(response_body);
+				return;
+			case http::StatusUnauthorized:
+				api_handler(expected::unexpected(MakeHTTPResponseError(
+					UnauthorizedError,
+					resp,
+					response_body,
+					"Failed to authorize with the server.")));
+				return;
+			case http::StatusBadRequest:
+			case http::StatusInternalServerError:
+				api_handler(expected::unexpected(MakeHTTPResponseError(
+					APIError, resp, response_body, "Failed to authorize with the server.")));
+				return;
+			default:
+				mlog::Error("Unexpected error code " + resp->GetStatusMessage());
+				api_handler(expected::unexpected(MakeError(
+					ResponseError, "Unexpected error code: " + resp->GetStatusMessage())));
```

Comment:
Ignoring `response_body` here.

Reply:
I only parse the body in the instances the API specifies we provide one, which is the sane thing to do imo

Reply:
I was thinking it could make sense to make it part of the error anyway. If it's empty, nothing is added, if it's non-empty, it might have valuable info.